skule_vote/backend/serializers.py

```diff
@@ -41,3 +43,54 @@ class Meta:
             "start_time",
             "end_time",
         )
+
+
+class BallotSerializer(serializers.Serializer):
+    electionId = serializers.CharField(max_length=200, allow_blank=False)
+    # rank -> candidate_id
+    ranking = serializers.DictField(allow_empty=True)
+
+    def validate(self, data):
+        """
+        Ensure election exists and candidates belong to the election.
+        """
+        if Election.objects.filter(id=data["electionId"]).count() != 1:
+            raise serializers.ValidationError(
+                "The election you are trying to vote in does not exist"
+            )
+
+        candidate_ids = [
+            c.id for c in Candidate.objects.filter(election=data["electionId"])
+        ]
+        for _, candidate in data["ranking"].items():
+            if candidate not in candidate_ids:
+                raise serializers.ValidationError(
+                    f"No candidate with id: {candidate} exists in election {data['electionId']}"
```

Comment:
You're not consistent with your text-to numbering here
either use `text: number` or `text number` for both instead of how you use `id: number` and `election number`